File: Cal/Al/src/C4/U4/AcyclicLP.cpp
#include "C4/U4/AcyclicLP.h"
#include "C1/Unit3_Stacks/Stack.h"

#include <limits>

AcyclicLP::AcyclicLP(EdgeWeightedDigraph* G, int s)
{
    edgeTo.reserve(G->getV());
    distTo.reserve(G->getV());

    for(int v = 0; v < G->getV(); v++)
    {
        edgeTo.push_back(nullptr);
        distTo[v] = new CDouble(numeric_limits<double>::min());
    }
    distTo[s]->setValue(0.0);

    Topological* top = new Topological(G);

    Stack<int>* order = top->getOrder();
    while(!order->isEmpty())
    {
        int v = order->pop();
        relax(G, v);
    }
}

void AcyclicLP::relax(EdgeWeightedDigraph* G, int v)
{
    Bag<DirectedEdge*> edgeBag;
    G->getAdj(v, edgeBag);
    Bag<DirectedEdge*>::Node* myB = edgeBag.first;
    while(nullptr != myB)
    {
        DirectedEdge* e = myB->item;

        int w = e->to();
        if(distTo[w]->getValue() < distTo[v]->getValue() + e->getWeight())
        {
            distTo[w]->setValue(distTo[v]->getValue() + e->getWeight());
            edgeTo[w] = e;
        }

        myB = myB->next;
    }
}

double AcyclicLP::getDistTo(int v)
{
    return distTo[v]->getValue();
}

bool AcyclicLP::hasPathTo(int v)
{
    return distTo[v]->getValue() < numeric_limits<double>::max();
}

void AcyclicLP::pathTo(int v, stack<DirectedEdge*>& path)
{
    if(!hasPathTo(v))
        return;
    for(DirectedEdge* e = edgeTo[v]; e != nullptr; e = edgeTo[e->from()])
        path.push(e);
}

File: Jal/Al/src/C4/U4/AcyclicLP.java
package C4.U4;

import C4.U2.Topological;
import C4.U3.Edge;
import C1.Unit3_Stacks.Stack;

import C4.U4.DirectedEdge;
import C4.U4.EdgeWeightedDigraph;
import StdLib.*;

public class AcyclicLP {
    private DirectedEdge[] edgeTo;
    private double[] distTo;

    public AcyclicLP(EdgeWeightedDigraph G, int s)
    {
        edgeTo = new DirectedEdge[G.V()];
        distTo = new double[G.V()];

        for(int v = 0; v < G.V(); v++)
            distTo[v] = Double.NEGATIVE_INFINITY;
        distTo[s] = 0.0;

        Topological top = new Topological(G);

        for(int v : top.order())
            relax(G, v);
    }

    private void relax(EdgeWeightedDigraph G, int v)
    {
        for(DirectedEdge e : G.adj(v))
        {
            int w = e.to();
            if(distTo[w] < distTo[v] + e.weight())
            {
                distTo[w] = distTo[v] + e.weight();
                edgeTo[w] = e;
            }
        }
    }

    public double distTo(int v)
    {
        return distTo[v];
    }

    public boolean hasPathTo(int v)
    {
        return distTo[v] < Double.POSITIVE_INFINITY;
    }

    public Iterable<DirectedEdge> pathTo(int v)
    {
        if(!hasPathTo(v))
            return null;
        Stack<DirectedEdge> path = new Stack<DirectedEdge>();
        for(DirectedEdge e = edgeTo[v]; e != null; e = edgeTo[e.from()])
            path.push(e);
        return path;
    }

    public static void main(String[] args)
    {
        EdgeWeightedDigraph G;
        G = new EdgeWeightedDigraph(new In(args[0]));
        int s = Integer.parseInt(args[1]);
        AcyclicLP sp = new AcyclicLP(G, s);

        for(int t = 0; t < G.V(); t++)
        {
            StdOut.print(s + " to " + t);
            StdOut.printf(" (%4.2f): ", sp.distTo(t));
            if(sp.hasPathTo(t))
                for(DirectedEdge e : sp.pathTo(t))
                    StdOut.print(e + "    ");
            StdOut.println();
        }
    }

}
